### PR TITLE
feat: [object-store] re-export hyper

### LINKED ISSUE #5386: [object-store] re-export `hyper`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

To be able to better handle errors of `object-store`, we'd like to downcast errors to `hyper` errors. For this to be possible the version of `hyper` must match that one used by `object-store`. 

Libraries can be dependent on different versions, so this is most robust and flexible with re-exporting `hyper`.

### PR DESCRIPTION
**Pull request by ritchie46**
Closes #5386

# Are there any user-facing changes?

No

### CODE REVIEW

File: object_store/src/export.rs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+//! Re-exports of libraries used by `object_store`.
```
Comment (tustvold) [inline object_store/src/export.rs:1]: You will need to add the license goop at the top to make CI happy

File: object_store/src/lib.rs [hunk 2]
```diff
@@ -497,6 +497,7 @@ pub use tags::TagSet;
 pub mod multipart;
 mod parse;
 mod util;
+pub mod export;
```
Comment (tustvold) [inline object_store/src/lib.rs:500]: Is this a convention found in other crates, or do they re-export at the root?

Comment (ritchie46) [reply object_store/src/lib.rs:500]: Could not really find a convention. In this book it is done in the root: https://www.lurklurk.org/effective-rust/re-export.html

Which should work fine for `hyper`, but might give you some collisions if a library has a name colliding with internal modules. No preference on my side. Shall I do it from root?

File: object_store/src/export.rs (hunk 1 above)
Comment (ritchie46) [reply object_store/src/export.rs:1]: Ah, right.

**Comment by tustvold** [comment]
So I did some digging because reqwest, which is the client we are actually using, doesn't export hyper and I wondered why. I _think_ it is something to do with the way that reqwest supports wasm32. This isn't an immediate issue, given wasm32 for HTTP stores is currently incomplete, but it does make me wonder if this is the correct way to go about this.

Perhaps something like #5345 might be what you are looking for?

**Comment by ritchie46** [comment]
> Perhaps something like https://github.com/apache/arrow-rs/issues/5345 might be what you are looking for?

Which is an object-store variant of hyper/http errors? Seems also good. Since `reqwest` doesn't export `hyper`, but you do use it for error handling. Do you ensure manually to keep the versions matched?

**Comment by tustvold** [comment]
> Do you ensure manually to keep the versions matched

Yes, and have tests that should catch any divergence. It is not an ideal situation, but it has worked so far

**Comment by ritchie46** [comment]
In that case, I would love to piggy back on that via an export. 😜

**Comment by tustvold** [comment]
Perhaps you might expand upon your use-case for downcasting the errors? It occurs to me that this is somewhat breaking encapsulation, and we might be able to expose this information in a better way, perhaps??

**Comment by ritchie46** [comment]
> Perhaps you might expand upon your use-case for downcasting the errors? It occurs to me that this is somewhat breaking encapsulation, and we might be able to expose this information in a better way, perhaps??

I want to get most information out of the errors and see how to handle upon those. I haven't got a clear answer to what exactly my solution is. But having access to the http errors for accessing s3 is very important to us as I want to be close to maximum concurrency without triggering those. We want to be very fast here and like to be able experiment with tuning settings. Ideally I want to dynamically tune the client settings later.

**Comment by ritchie46** [comment]
> Because there is no requirement that all or even any ObjectStore are implemented using hyper,

I totally understand that. And I as a user am willing to pay that breaking change risk. ;)

So we could `export` it under an `unstable` flag. Or if you don't want to expose it in `object-store` also fine, then we copy the `hyper` version from `object-store` and do it manually. Either way, I want to tinker. ^^

> Perhaps you could identify from the list of errors [exposed by S3](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#RESTErrorResponses) what variant you are interested in intercepting.

I don't know yet. I always learn while tinkering. :sweat_smile:. It isn't a joke, I cannot answer what I need exactly yet. Have to learn this as we go.

If we have found something that works good for us, we can push that back into object-store. I rather have that handled higher up.

**Comment by tustvold** [comment]
In that case, how about we mark this PR as a draft, and revisit this once we have a better idea of the concrete use-cases for this feature?

**Comment by ritchie46** [comment]
Sure, fine by me. :+1:

**Pull request closed without merging**